website/mailing_list/utils.py

```diff
@@ -0,0 +1,422 @@
+    if subscription:
+        subscription.remove_user_from_subscription(user)
+        node.mailing_updated = True
+        node.save()
+    celery_remove_user_from_list(node._id, user._id)
+
+
+@user_confirmed.connect
+def resubscribe_on_confirm(user):
+    for node in user.contributed:
+        subscribe_contributor_to_mailing_list(node, user)
+
+@node_deleted.connect
+def remove_list_for_deleted_node(node):
+    node.mailing_updated = True
+    node.save()
+    celery_delete_list(node._id)
+
+
+###############################################################################
+# Mailing List Helper Functions
+###############################################################################
+
+def jsonify_users_list(users, unsubs=[]):
+    """ Serializes list of users for Mailgun
+
+    :param list users: users to serialize
+    :param list unsubs: unsubscribed users
+    """
+    members_list = []
+    for member in users:
+        for email in member.emails:
+            members_list.append({
+                'address': email,
+                'subscribed': member not in unsubs and email == member.username,
+                'vars': {'_id': member._id, 'primary': email == member.username}
+            })
+    return members_list
+
+def address(node_id):
+    return '{}@{}'.format(node_id, furl(settings.DOMAIN).host)
+
+def list_title(node):
+    return '{} Mailing List'.format(node.title)
+
+def find_email(long_email):
+    # allow for both "{email}" syntax and "{name} <{email}>" syntax
+    if '<' in long_email:
+        email_match = ANGLE_BRACKETS_REGEX.search(long_email)
+        if email_match:
+            return email_match.groups()[0].lower().strip()
+    elif '@' in long_email:
+        return long_email.lower().strip()
+    return None
+
+def log_message(**kwargs):
+    """ Acquires and logs messages sent through Mailgun"""
+    from website.models import Node  # avoid circular imports
+    message = request.form
+    target = find_email(message['To'])
+    node = Node.load(re.search(r'[a-z0-9]*@', target).group(0)[:-1])
+
+    sender_email = find_email(message['From'])
+    sender = get_user(email=sender_email)
+
+    # Create a log of this mailing event
+    MailingListEventLog(
+        email_content=message,
+        destination_node=node,
+        sending_user=sender,
+    ).save()
+
+def unsubscribe_user_hook(*args, **kwargs):
+    """ Hook triggered by MailGun when user unsubscribes.
+    See `Unsubscribes Webhook` below https://documentation.mailgun.com/user_manual.html#tracking-unsubscribes
+    for possible kwargs
+    """
+    message = request.form
+    unsub = message.get('recipient')
+    mailing_list = message.get('mailing-list')
+    if not unsub or not mailing_list:
+        raise Exception()
+    from website.models import User, NotificationSubscription  # avoid circular imports
+    user = User.find_one('username', 'eq', unsub)
```

Comment:
This hook should handle the case where no user matches the query. `find_one` will raise a `NoResultsFound` exception.